src/eda_server/ruleset.py

```diff
@@ -73,13 +78,194 @@ def ensure_directory(directory):
+                "ExtraHosts": ["host.docker.internal:host-gateway"],
+                "ExposedPorts": {"8000/tcp": {}},
+                "HostConfig": {
+                    "PortBindings": {"8000/tcp": [{"HostPort": "8000"}]},
+                    "NetworkMode": "eda-network",
+                },
+            }
+        )
+        logger.debug("Starting fallback/default container")
+        await container.start()
+    except aiodocker.exceptions.DockerError as e:
+        logger.error("Failed to start fallback/default container: %s", e)
+        await container.delete()
+        await docker.close()
+        raise
+
+    return (docker, container)
+
+
+async def create_websocket_docker_activation(
+    activation_id: int,
+    source: Dict,
+    execution_environment: str,
+    default_port: Union[str, int],
+    default_host: str = DEFAULT_HOST,
+) -> Tuple[aiodocker.docker.Docker, aiodocker.docker.DockerContainer]:
+    config = source.get("config", {})
+    host = config.get("host", default_host)
+    port = config.get("port", default_port)
+    exposed_ports = config.get("exposed_ports", {"8000/tcp": {}})
+    extra_hosts = config.get(
+        "extra_hosts", ["host.docker.internal:host-gateway"]
+    )
+    host_config = config.get(
+        "host_config",
+        {
+            "PortBindings": {"8000/tcp": [{"HostPort": "8000"}]},
+            "NetworkMode": "eda-network",
+        },
+    )
+    environment = config.get("environment", ["ANSIBLE_FORCE_COLOR=True"])
+
+    docker = aiodocker.Docker()
+    try:
+        await aiodocker.images.DockerImages(docker).pull(execution_environment)
+
+        logger.debug("Creating container")
+        logger.debug("Host: %s", host)
+        logger.debug("Port: %s", port)
+        container = await docker.containers.create(
+            {
+                "Cmd": [
+                    "ssh-agent",
+                    "ansible-rulebook",
+                    "--worker",
+                    "--websocket-address",
+                    f"ws://{host}:{port}/api/ws2",
+                    "--id",
+                    str(activation_id),
+                    "--debug",
+                ],
+                "Image": execution_environment,
+                "Env": environment,
+                "ExtraHosts": extra_hosts,
+                "ExposedPorts": exposed_ports,
+                "HostConfig": host_config,
+            }
+        )
+        logger.debug("Starting container")
+        await container.start()
+    except aiodocker.exceptions.DockerError as e:
+        logger.error("Failed to start container: %s", e)
+        await container.delete()
+        await docker.close()
+        raise
+
+    return (docker, container)
+
+
+def resolve_activation_function(source: Dict) -> Callable:
+    # TODO: React to source type
+    activation_type = "websocket"
+    activation_func_name = f"create_{activation_type}_docker_activation"
+    activation_func = globals()[activation_func_name]
```

Comment:
Please don't do that. If you need to resolve a function by it's name, create a dictionary of names and references to functions.